common/src/main/java/com/microsoft/identity/common/internal/result/MsalBrokerResultAdapter.java

```diff
@@ -182,12 +183,112 @@ public BaseException getBaseExceptionFromBundle(@NonNull final Bundle resultBund
+        final String exceptionType = brokerResult.getExceptionType();
 
+        if(!TextUtils.isEmpty(exceptionType)){
+            return getBaseExceptionFromExceptionType(exceptionType, brokerResult);
+        }else {
+            // This code is here for legacy purposes where old versions of broker (3.1.8 or below)
+            // wouldn't return exception type in the result.
+            Logger.info(TAG, "Exception type is not returned from the broker, " +
+                    "using error codes to transform to the right exception");
+            return getBaseExceptionFromErrorCodes(brokerResult);
+        }
+
+    }
+
+    private BaseException getBaseExceptionFromExceptionType(@NonNull final String exceptionType,
+                                                            @NonNull final BrokerResult brokerResult){
+        BaseException baseException = null;
+
+        if(exceptionType.equalsIgnoreCase(UiRequiredException.sName)){
+            Logger.warn(TAG, "Received a UIRequired exception from Broker : "
+                    + brokerResult.getErrorCode()
+            );
+            baseException = new UiRequiredException(
+                    brokerResult.getErrorCode(),
+                    brokerResult.getErrorMessage()
+            );
+        }
+
+        if(exceptionType.equalsIgnoreCase(ServiceException.sName)){
+            Logger.warn(TAG, "Received a Service exception from Broker : "
+                    + brokerResult.getErrorCode()
+            );
+            baseException = getServiceException(brokerResult);
+        }
+
+        if(exceptionType.equalsIgnoreCase(IntuneAppProtectionPolicyRequiredException.sName)){
+            Logger.warn(TAG,
+                    "Received a IntuneAppProtectionPolicyRequiredException exception from Broker: "
+                    + brokerResult.getErrorCode()
+            );
+            baseException = getIntuneProtectionRequiredException(brokerResult);
+        }
+
+        if(exceptionType.equalsIgnoreCase(UserCancelException.sName)){
+            Logger.warn(TAG, "Received a User cancelled exception from Broker : "
+                    + brokerResult.getErrorCode()
+            );
+            baseException = new UserCancelException();
+        }
+
+        if(exceptionType.equalsIgnoreCase(ClientException.sName)){
+            Logger.warn(TAG, "Received a ClientException exception from Broker : "
+                    + brokerResult.getErrorCode()
+            );
+            baseException = new ClientException(
+                    brokerResult.getErrorCode(),
+                    brokerResult.getErrorMessage()
+            );
+        }
+
+        if(exceptionType.equalsIgnoreCase(ArgumentException.sName)){
+            Logger.warn(TAG, "Received a Argument exception from Broker : "
+                    + brokerResult.getErrorCode()
+            );
+            baseException = new ArgumentException(
+                    ArgumentException.ACQUIRE_TOKEN_OPERATION_NAME,
```

Comment:
How do we know that it is `ACQUIRE_TOKEN_OPERATION_NAME`?

Reply:
Good point, I introduced a new operation to indicate that it'a broker request

Reply:
What about silent?

Reply:
`ACQUIRE_TOKEN_SILENT_OPERATION_NAME` ?

Reply:
I have introduced a new operation name `BROKER_TOKEN_REQUEST_OPERATION_NAME` which is passed here, since the basic MSAL argument validation is successful and is sent to broker, here if an argument exception is thrown it would be of broker. For now I added a generic operation name here, if need specific things in the future it can be added as needed